File: Binary_Search.java
// Binary search algorithm always need the sorted list of the data
// Initialy, this code is written for the ascending sorted data. So, must of the comments referes to the ascending data
// But, hint is given in the for descending sorted data.

public class Binary_Search{
    public static int search( int []list, int high, int low, int searchValue) {
        if (low <= high){
            // calculate the mid value 
            int mid = (high+low )/2;      // decimal value is automatically round up

            // return mid, if search value in the mid index by itself
            if( list[mid] == searchValue){
                return mid;
            }
            // if searchValue is greater than the mid value
            // that means search value is loacted right side of the mid value
            // So, low is need to update with mid+1 
            //## recall the search function with updated low value
            // other value is remain same as before
            else if ( list[mid] < searchValue){         // change:  list[mid] > searchValue    // descending sorted data
                return search(list, high, mid+1,searchValue);
            }
            else{
                //  if the search value is smaller than the mid value
                // that means search value is located left side of the mid value 
                // So, high is need to upadate with mid-1
                //## recall the search function with updated high value,
                // other value is remain same as before
                return search(list,mid-1,low,  searchValue);
            }
        }
        // return the value -1 if the search value is not found
        return -1;
    }
    public static void main(String[] args) {
        // creating the list
        int a[] = {1,2,3,4,5,6,7};          // ascending sorted data
        // int a[] = {9,8,7,6,5,4,3,2,1};           // descending sorted data
        int low = 0;                        // first index of the list
        int high = a.length-1;      // last index of the list
        int searchValue = 6;        // search value
        int output = search(a, high, low, searchValue); 
        if(output == -1){       // if value is not found
            System.out.println("Value not found!!");
        }
        else{
            System.out.println("The searched value is : "+ a[output]+" found at index: "+output);
        }
    }
}

File: BinarySearch.py
#  Binary search algorithm
#  Binary search algorithm always needs the sorted lists of data
#  this code only works on the ascending data i.e. [1,2,3,4,5,6,7,8]

def search(List, lowIndex, highIndex, searchValue):
    #  lower index must be smaller than the higher 
    if(lowIndex <= highIndex):

        # calculaing the mid index of the List
        midIndex = (lowIndex + highIndex) //2

        # if search value is in middle index by itself
        if List[midIndex] == searchValue:
            return midIndex
        
        # if the mid index value is smaller than the search value
        # then, search value is located in the right side of the mid index
        elif List[midIndex] <  searchValue:             # change: List[midIndex] > searchValue      # for desceding sorted values
            return search(List, midIndex+1, highIndex, searchValue)

        # Else when mid index value is greater than the seach value
        # then, search value is located left side of the mid index
        else:
            return search(List,lowIndex, midIndex -1, searchValue)
    
    # if value is not in the list it returns -1 
    else:
        return -1

# creating required parameters for search function
numbers = [1,2,3,4,5,6,7,8,9]
# numbers = [9,8,7,6,5,4,3,2,1]
high = len(numbers)-1
low = 0
search_value = 7

# calling the search function
result = search(numbers, low, high, search_value)

# displaying the search result
if result == -1:
    print("Value is not found!!!")
else:
    print("The value is at index of %d " %result)